### PR TITLE
Custom renderer

### PR DESCRIPTION
**Pull request by github-richard-depierre**
We are introducing the PluginRenderer as we did on [android](https://docs.prebid.org/prebid-mobile/pbm-api/android/pbm-plugin-renderer.html).

### CODE REVIEW

File: PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRenderer.swift [hunk 1]
```diff
@@ -0,0 +1,53 @@
+/*   Copyright 2018-2021 Prebid.org, Inc.
```
Comment (github-lucas-bon) [inline PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRenderer.swift:1]: PrebidMobile Classes Diagrams (1)

We have this class diagram about how these classes suppose to be in terms of naming and declaration. 
To check:
- We do we have to keep consistency between iOS and Android naming? I see differences we are introducing between the OS's, but perhaps it should keep like this? 
Such a symmetry may not be in place already, see that in Android project the class do register the plugin renderer is name PrebidMobile and in iOS project Prebid.

File: PrebidMobile/PrebidMobileRendering/PluginRenderer/PluginEventListener.swift [hunk 2]
```diff
@@ -0,0 +1,19 @@
+/*   Copyright 2018-2021 Prebid.org, Inc.
```
Comment (github-lucas-bon) [inline PrebidMobile/PrebidMobileRendering/PluginRenderer/PluginEventListener.swift:1]: Screenshot 2024-05-16 at 10 51 37

This class supposed to require for those implementing it the following field `pluginName: String`

File: PrebidMobile/ConfigurationAndTargeting/Prebid.swift [hunk 3]
```diff
@@ -225,4 +225,12 @@ public class Prebid: NSObject {
     override init() {
         timeoutMillis = defaultTimeoutMillis
     }
+    
+    public static func registerPluginRenderer(_ prebidMobilePluginRenderer: PrebidMobilePluginRenderer) {
```
Comment (github-lucas-bon) [inline PrebidMobile/ConfigurationAndTargeting/Prebid.swift:229]: Screenshot 2024-05-16 at 10 53 54

It Supposed to comply with the class diagram, is it missing a function to check if a plugin renderer exists

File: PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRegister.swift [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+ */
+
+import Foundation
+
+/// Global singleton responsible to store plugin renderer instances
+@objc public class PrebidMobilePluginRegister: NSObject {
+    @objc public static let shared = PrebidMobilePluginRegister()
+    
+    private let queue = DispatchQueue(label: "PrebidMobilePluginRegisterQueue", attributes: .concurrent)
+    private var plugins = [String: PrebidMobilePluginRenderer]()
```
Comment (github-lucas-bon) [inline PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRegister.swift:23]: There is no way to set it as let and the type a sort of mutable map?

File: PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRegister.swift [hunk 5]
```diff
@@ -0,0 +1,91 @@
+/*   Copyright 2018-2021 Prebid.org, Inc.
```
Comment (github-lucas-bon) [inline PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRegister.swift:1]: Screenshot 2024-05-16 at 11 05 18

Some things don't match with the class diagram:

- No function for unregistering event listener and containsPlugin
- getRTBListOfRenderers suppose to return a list of PrebidMobilePluginRenderer

File: PrebidMobile/PrebidMobileRendering/Prebid/PBMCacheRenderers/PBMDisplayView.m [hunk 6]
```diff
@@ -70,9 +70,24 @@ - (void)displayAd {
...
+    [self.renderer setupBid:self.bid
+            adConfiguration:self.adConfiguration
+                 connection:self.connection ?: PrebidServerConnection.shared
+                     callback:^(PBMTransaction * _Nullable transaction,
+                                NSError * _Nullable error) {
+        @strongify(self);
+        if (error) {
+            [self reportFailureWithError:error];
+        } else {
+            //[self reportSuccess];
```
Comment (github-lucas-bon) [inline PrebidMobile/PrebidMobileRendering/Prebid/PBMCacheRenderers/PBMDisplayView.m:87]: Please clarity this `//[self reportSuccess];`

File: PrebidMobile/PrebidMobileRendering/Prebid/PBMCacheRenderers/PrebidRenderer.swift [hunk 7]
```diff
@@ -0,0 +1,34 @@
+/*   Copyright 2018-2021 Prebid.org, Inc.
```
Comment (github-lucas-bon) [inline PrebidMobile/PrebidMobileRendering/Prebid/PBMCacheRenderers/PrebidRenderer.swift:1]: It supposed to implement interstitial loading function as well.

File: tools/PrebidValidator/Dr.Prebid/TestSummary/AdServerValidation/PBVLineItemsSetupValidator.m [hunk 8]
```diff
@@ -89,13 +89,13 @@ - (void)startTest
     NSString *adUnitID = [[NSUserDefaults standardUserDefaults] stringForKey:kAdUnitIdKey];
     NSString *bidPrice = [[NSUserDefaults standardUserDefaults] stringForKey:kBidPriceKey];
         
-    GADAdSize GADAdSize = kGADAdSizeInvalid;
+    GADAdSize GADAdSize = GADAdSizeInvalid;
```
Comment (github-lucas-bon) [inline tools/PrebidValidator/Dr.Prebid/TestSummary/AdServerValidation/PBVLineItemsSetupValidator.m:92]: Clarify why we have the changes in the file

**Review by github-lucas-bon — COMMENTED** [review]
Hello @github-richard-depierre 

I added some considerations in some files and here I am going to provide some more general insights.

I recomend you to check the docs and diagrams we have for the Plugin renderer, let me know if you need help to understand more.

Lucid board: https://lucid.app/lucidchart/804af2fe-c369-4272-b6fa-d14fd614819c/edit?viewport_loc=-2544%2C-3354%2C6944%2C6011%2C0_0&invitationId=inv_71171ef5-d628-47ef-ae6e-3e48705adacc

PDF:
[Prebid Renderer Diagrams.pdf](https://github.com/prebid/prebid-mobile-ios/files/15332645/Prebid.Renderer.Diagrams.pdf)

Android docs:
https://docs.prebid.org/prebid-mobile/pbm-api/android/pbm-plugin-renderer.html

- The default prebid renderer supposed to be registered during the SDK initialisation [ref](https://github.com/prebid/prebid-mobile-android/blob/c12a2747243dcc65cfe1ea374571ceee43d22b4c/PrebidMobile/PrebidMobile-core/src/main/java/org/prebid/mobile/rendering/sdk/SdkInitializer.java#L56)
- In order to start to see if the implementation is working as expected we should start to implement some unit tests for the general behavior from the plugin renderer registering 
- We should add some use cases tests as well [as we did for Android](https://docs.prebid.org/prebid-mobile/pbm-api/android/pbm-plugin-renderer.html#resources)
- Let's improve as well the description of this pull request, since this is an open source project as much as context we add more easy we will turn the life of those reviewing it

So that is it for now, I will try to suggest more fixes and improvements in the next pull requests.

File: tools/PrebidValidator/Dr.Prebid/TestSummary/AdServerValidation/PBVLineItemsSetupValidator.m (hunk 8 above)
Comment (github-richard-depierre) [reply tools/PrebidValidator/Dr.Prebid/TestSummary/AdServerValidation/PBVLineItemsSetupValidator.m:92]: This is because it has been changed by GAD and the name has changed now.

File: InternalTestApp/PrebidMobileDemoRendering/ViewControllers/AboutViewController.swift [hunk 9]
```diff
@@ -33,7 +33,7 @@ class AboutViewController : FormViewController {
             "OMSDK_Prebidorg": Prebid.shared.omsdkVersion
         ]),
         ("Ad Server SDKs", [
-            "GoogleMobileAds SDK": GADMobileAds.sharedInstance().sdkVersion,
+            "GoogleMobileAds SDK": GADMobileAds.sharedInstance().versionNumber,
```
Comment (github-lucas-bon) [inline InternalTestApp/PrebidMobileDemoRendering/ViewControllers/AboutViewController.swift:36]: It this causing building issues?

File: PrebidMobile/PrebidMobileRendering/ORTB/PBMORTBRendererConfig.m..m [hunk 10]
```diff
@@ -0,0 +1,22 @@
+//
```
Comment (github-lucas-bon) [inline PrebidMobile/PrebidMobileRendering/ORTB/PBMORTBRendererConfig.m..m:1]: I guess there is an issue with the extension here?
PrebidMobile/PrebidMobileRendering/ORTB/PBMORTBRendererConfig`.m..m`

File: PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRegister.swift [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+        queue.async(flags: .barrier) { [plugins] in
+            plugins
+                .values
+                .forEach {
+                    $0.unregisterEventDelegate?(pluginEventDelegate: pluginEventDelegate, adUnitConfigFingerprint: adUnitConfigFingerprint)
+                }
+        }
+    }
+    
+//    /// Returns the list of available renderers for the given ad unit for RTB request -> What about interstitial ?
```
Comment (github-lucas-bon) [inline PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRegister.swift:85]: Removable comments?

File: PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRenderer.swift [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+import Foundation
+
+@objc public protocol PrebidMobilePluginRenderer: AnyObject {
+    
+    @objc var name: String { get }
+    @objc var version: String { get }
+    @objc var token: String? { get }
+
+    /// Returns true only if the given ad unit could be renderer by the plugin
+    @objc func isSupportRendering(for format: AdFormat?) -> Bool
```
[the remark this answers is not in the document]
Comment (github-lucas-bon) [reply PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRenderer.swift:33]: @YuriyVelichkoPI WDYT?

File: InternalTestApp/PrebidMobileDemoRendering/ViewControllers/AboutViewController.swift (hunk 9 above)
Comment (github-richard-depierre) [reply InternalTestApp/PrebidMobileDemoRendering/ViewControllers/AboutViewController.swift:36]: Yes it is, we can't build otherwise.

**Review by github-lucas-bon — COMMENTED** [review]
I would suggest next steps:

For both points below take into account these json samples, you can as well inspect the network with proxyman using prebid sample app https://teads.atlassian.net/wiki/spaces/INAPP/pages/6547931197/Resources

- Ensure the bid request builder is adding your registered plugin renderers in the ad request. `getRTBListOfRenderersFor` being called to fill the existed plugin renderers in the bid request?
- Ensure the bid response is being parsed taking into account the present plugin renderers in the response (name and version)
- Resolve conflicts with master

-- If you achieve the points below I believe we could request a review from Prebid team.

___

Points above could be achieved with an implementation on sample app. 
(Could it be added in the same PR?)

- Ensure that you init the default Prebid render in the initialisation of the app
- Ensure the in the absence of other plugin renderers, the default one will be always take ownership about the ad rendering
- Create a sample plugin renderer for testing purposes
- Ensure your sample plugin renderer is sent in the bid request, parsed in the bid response and correctly delegated for rendering 
- Ensure you can dispatch custom event listener from your plugin renderer

File: PrebidMobile/PrebidMobileRendering/ORTB/PBMORTBRendererConfig.h [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+  */
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface PBMORTBRendererConfig : NSObject
```
Comment (github-paul-nicolas) [inline PrebidMobile/PrebidMobileRendering/ORTB/PBMORTBRendererConfig.h:20]:
```suggestion
@interface PBMORTBRendererConfig : PBMORTBAbstract
```

File: InternalTestApp/PrebidMobileDemoRendering/ViewControllers/AboutViewController.swift [hunk 14]
```diff
@@ -68,7 +68,7 @@ class AboutViewController : FormViewController {
             for (frameworkName, frameworkVersion) in frameworks {
                 section <<< LabelRow() { row in
                     row.title = frameworkName
-                    row.value = frameworkVersion
+                    row.value = frameworkVersion as? Cell<String>.Value
```
Comment (github-richard-depierre) [inline InternalTestApp/PrebidMobileDemoRendering/ViewControllers/AboutViewController.swift:71]:
```suggestion
                    row.value = frameworkVersion as? Cell<String>.Value
```
```suggestion
                    row.value = frameworkVersion as? Cell<String>.Value
```

File: InternalTestApp/PrebidMobileDemoRendering/ViewControllers/Adapters/Prebid/CustomRenderer/CustomRendererBannerController.swift [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+        
+        reloadButton.addTarget(self, action: #selector(reload), for: .touchUpInside)
+        stopRefreshButton.addTarget(self, action: #selector(stopRefresh), for: .touchUpInside)
+        
+        setupAdapterController()
+    }
+    
+    deinit {
+        Targeting.shared.sourceapp = nil
+        Prebid.unregisterPluginRenderer(SampleCustomRenderer())
```
Comment (github-lucas-bon) [inline InternalTestApp/PrebidMobileDemoRendering/ViewControllers/Adapters/Prebid/CustomRenderer/CustomRendererBannerController.swift:59]: I think it is more consistent if you init SampleCustomRenderer() instance in a const instead of passing always new instances. If helps the implementation to be future proof. 
For instance `unregisterPluginRenderer` might be changed and take into account not only name but the instance register.

le sampleCustomRenderer = SampleCustomRenderer() 
Prebid.registerPluginRenderer(sampleCustomRenderer)
Prebid.unregisterPluginRenderer(sampleCustomRenderer)

File: InternalTestApp/PrebidMobileDemoRendering/ViewControllers/AboutViewController.swift (hunk 14 above)
Comment (github-lucas-bon) [inline InternalTestApp/PrebidMobileDemoRendering/ViewControllers/AboutViewController.swift:71]: Why this change was added?

File: InternalTestApp/PrebidMobileDemoRendering/ViewControllers/Adapters/Prebid/CustomRenderer/CustomRendererInterstitialController.swift [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+        self.adapterViewController = rootController
+        Prebid.registerPluginRenderer(SampleCustomRenderer())
+        super.init()
+        
+        setupAdapterController()
+    }
+    
+    deinit {
+        Targeting.shared.sourceapp = nil
+        Prebid.unregisterPluginRenderer(SampleCustomRenderer())
```
Comment (github-lucas-bon) [inline InternalTestApp/PrebidMobileDemoRendering/ViewControllers/Adapters/Prebid/CustomRenderer/CustomRendererInterstitialController.swift:59]: Same consideration left for the banner controller

File: PrebidMobile/PrebidMobileRendering/ORTB/PBMORTBBidRequest.m [hunk 17]
```diff
@@ -37,6 +38,7 @@ - (nonnull instancetype)init {
     _user = [PBMORTBUser new];
     _regs = [PBMORTBRegs new];
     _source = [PBMORTBSource new];
+    _data = [PBMORTBData new];
```
Comment (github-lucas-bon) [inline PrebidMobile/PrebidMobileRendering/ORTB/PBMORTBBidRequest.m:41]: Double check the utility for this new field

File: PrebidMobile/PrebidMobileRendering/ORTB/PBMORTBData.m [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+  */
+
+#import "PBMORTBData.h"
+#import "PBMORTBAbstract+Protected.h"
+
+@implementation PBMORTBData
```
Comment (github-lucas-bon) [inline PrebidMobile/PrebidMobileRendering/ORTB/PBMORTBData.m:19]: Is this being used?

File: PrebidMobile/PrebidMobileRendering/Prebid/PBMCacheRenderers/PBMDisplayView.m [hunk 19]
```diff
@@ -208,4 +169,8 @@ - (void)interactionDelegateDidDismissModal {
     }
 }
 
+- (void) adViewLoaded:(UIView *)adView adSize:(CGSize)adSize {
+    // needed for custom renderers
```
Comment (github-lucas-bon) [inline PrebidMobile/PrebidMobileRendering/Prebid/PBMCacheRenderers/PBMDisplayView.m:173]: Why this is needed?
